understory_index/src/backends/grid.rs

```diff
@@ -0,0 +1,390 @@
+    }
+
+    fn covered_cells(&self, aabb: &Aabb2D<T>) -> SmallVec<[(i32, i32); 4]> {
+        let (ix0, ix1) = self.cell_range(aabb.min_x, aabb.max_x, self.origin_x);
+        let (iy0, iy1) = self.cell_range(aabb.min_y, aabb.max_y, self.origin_y);
+        let mut out: SmallVec<[(i32, i32); 4]> = SmallVec::new();
+        for ix in ix0..=ix1 {
+            for iy in iy0..=iy1 {
+                out.push((ix, iy));
+            }
+        }
+        out
+    }
+}
+
+impl<T: GridScalar> Backend<T> for Grid<T> {
+    fn insert(&mut self, slot: usize, aabb: Aabb2D<T>) {
+        self.ensure_slot(slot);
+
+        // If this slot was previously used, clean up its old cell memberships.
+        if let Some(old) = self.slots[slot].take() {
+            self.remove_from_cells(slot, &old.cells);
+        }
+
+        let cells = self.covered_cells(&aabb);
+        for &(ix, iy) in &cells {
+            self.cells.entry((ix, iy)).or_default().slots.push(slot);
+        }
+        self.slots[slot] = Some(SlotEntry { aabb, cells });
+    }
+
+    fn update(&mut self, slot: usize, aabb: Aabb2D<T>) {
+        // Take the current entry out to avoid aliasing `self` while mutating
+        // grid cells.
+        let current = if let Some(slot_ref) = self.slots.get_mut(slot) {
+            slot_ref.take()
+        } else {
+            None
+        };
+
+        let Some(mut entry) = current else {
+            // If the slot does not exist, treat this as an insert.
+            self.insert(slot, aabb);
+            return;
+        };
+
+        // If the AABB is unchanged, restore the entry and skip work.
+        if entry.aabb == aabb {
+            self.slots[slot] = Some(entry);
+            return;
+        }
+
+        // Remove from old cells.
+        self.remove_from_cells(slot, &entry.cells);
+
+        // Insert into new cells.
+        let cells = self.covered_cells(&aabb);
+        for &(ix, iy) in &cells {
+            self.cells.entry((ix, iy)).or_default().slots.push(slot);
+        }
+        entry.aabb = aabb;
+        entry.cells = cells;
+        self.slots[slot] = Some(entry);
+    }
+
+    fn remove(&mut self, slot: usize) {
+        if slot >= self.slots.len() {
+            return;
+        }
+        if let Some(entry) = self.slots[slot].take() {
+            self.remove_from_cells(slot, &entry.cells);
+        }
+    }
+
+    fn clear(&mut self) {
+        self.cells.clear();
+        self.slots.clear();
+    }
+
+    fn visit_point<F: FnMut(usize)>(&self, x: T, y: T, mut f: F) {
+        let ix = T::cell_coord(x, self.origin_x, self.cell_size);
+        let iy = T::cell_coord(y, self.origin_y, self.cell_size);
+        if let Some(cell) = self.cells.get(&(ix, iy)) {
+            for &slot in &cell.slots {
+                if let Some(Some(entry)) = self.slots.get(slot)
+                    && entry.aabb.contains_point(x, y)
+                {
+                    f(slot);
+                }
```

Comment:
I think we should be able to write `let entry = &self.slots[slot]` here (if I'm understanding right that if the slot isn't present, that's a serious bug in the data structure).